### PR TITLE
fix(python): Fix `row_by_key` typing

### PR DESCRIPTION
**Pull request by gab23r**
fixes: #19887

### CODE REVIEW

**Comment by gab23r** [comment]
I cant find why mypy complains... 
@alexander-beedie could you help me out ?

**Comment by hamdanal** [comment]
Thanks for fixing this, I am having the same problem in my code.

This function should have 4 overloads as it branches on both `named` and `unique` arguments:
- If `unique=False` (default), we get a list of objects (type unknown statically), if it is `True`, we get one object (or a tuple, depending on the number of columns)
- If `named=False` (default), we get an unknown object, if it is `True`, unknown objects get wrapped in a dict with string keys

Your current implementation only considers one combination out of four. The full signature should be this:
```py
    @overload
    def rows_by_key(
        self,
        key: ColumnNameOrSelector | Sequence[ColumnNameOrSelector],
        *,
        named: Literal[False] = False,
        include_key: bool = False,
        unique: Literal[False] = False,
    ) -> dict[Any, list[Any]]: ...
    @overload
    def rows_by_key(
        self,
        key: ColumnNameOrSelector | Sequence[ColumnNameOrSelector],
        *,
        named: Literal[False] = False,
        include_key: bool = False,
        unique: Literal[True],
    ) -> dict[Any, Any]: ...
    @overload
    def rows_by_key(
        self,
        key: ColumnNameOrSelector | Sequence[ColumnNameOrSelector],
        *,
        named: Literal[True],
        include_key: bool = False,
        unique: Literal[False] = False,
    ) -> dict[Any, list[dict[str, Any]]]: ...
    @overload
    def rows_by_key(
        self,
        key: ColumnNameOrSelector | Sequence[ColumnNameOrSelector],
        *,
        named: Literal[True],
        include_key: bool = False,
        unique: Literal[True],
    ) -> dict[Any, dict[str, Any]]: ...
    def rows_by_key(
        self,
        key: ColumnNameOrSelector | Sequence[ColumnNameOrSelector],
        *,
        named: bool = False,
        include_key: bool = False,
        unique: bool = False,
    ) -> dict[Any, Any]:
```

**Comment by gab23r** [comment]
Thanks hamdanal, I modified the signature based on your implementation and fixed the mypy error.

File: py-polars/polars/dataframe/frame.py [hunk 1]
```diff
@@ -10349,14 +10349,55 @@ def rows(
...
 
+    @overload
+    def rows_by_key(
+        self,
+        key: ColumnNameOrSelector | Sequence[ColumnNameOrSelector],
+        *,
+        named: Literal[False] = ...,
+        include_key: bool = ...,
+        unique: Literal[False] = ...,
+    ) -> dict[Any, Iterable[tuple[Any, ...]]]: ...
```
Comment (hamdanal) [inline py-polars/polars/dataframe/frame.py:10360]: This is not correct, the values of the dictionary are not iterables of tuple, they should be a "list of Any":
```py
import polars as pl
df = pl.DataFrame({
       "w": ["a", "b"],
       "x": ["q", "q"]
})
print(repr(df.rows_by_key(key=["w"])))
# defaultdict(<class 'list'>, {'a': ['q'], 'b': ['q']})
```
Actually all return types here are not correct. See my previous comment on this PR for the correct types.

Comment (gab23r) [reply py-polars/polars/dataframe/frame.py:10360]: arff, the type depends on the number of columns, see:

```python
import polars as pl
df = pl.DataFrame({
       "w": ["a", "b"],
       "x": ["q", "q"],
       "z": ["v", "v"]
})
print(repr(df.rows_by_key(key=["w"], named=False, unique=True)))
# {'a': ('q', 'v'), 'b': ('q', 'v')}
```
But in any case, you are right, this is not correct.

**Comment by alexander-beedie** [comment]
> I cant find why mypy complains... @alexander-beedie could you help me out ?

Looks like I have been beaten to the punch - thanks @hamdanal  ;)

**Review by alexander-beedie — APPROVED** [review]

**Pull request merged**